File: kat/src/lib/event-stream.ts
// Copyright © 2021 Kaleido, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as QueryString from 'querystring';
import { URL } from 'url';
import axios, { AxiosInstance } from 'axios';
import { promisify } from 'util';
import * as timers from 'timers';
import * as database from '../clients/database';
const sleep = promisify(timers.setTimeout);

import * as utils from './utils';
import { config } from './config'
import { IConfig, IEventStream, IEventStreamSubscription } from './interfaces';
const logger = utils.getLogger('lib/event-stream.ts');

/* istanbul ignore next */
const requestLogger = (config: any) => {
  const qs = config.params ? `?${QueryString.stringify(config.params)}` : '';
  logger.info(`--> ${config.method} ${config.baseURL}${config.url}${qs}`);
  logger.debug(config.data);
  return config;
};

/* istanbul ignore next */
const responseLogger = (response: any) => {
  const { config, status, data } = response;
  logger.info(`<-- ${config.method} ${config.url} [${status}]`);
  logger.debug(data);
  return response;
};

/* istanbul ignore next */
const errorLogger = (err: any) => {
  const { config = {}, response = {} } = err;
  const { status, data } = response;
  logger.info(`<-- ${config.method} ${config.url} [${status || err}]: ${JSON.stringify(data)}`);
  throw err;
};

const subscriptionsInfoEthereum = [
  ['Asset instance created', 'AssetInstanceCreated'],
  ['Asset instance batch created', 'AssetInstanceBatchCreated'],
  ['Payment instance created', 'PaymentInstanceCreated'],
  ['Payment definition created', 'PaymentDefinitionCreated'],
  ['Asset definition created', 'AssetDefinitionCreated'],
  ['Asset instance property set', 'AssetInstancePropertySet'],
  ['Described payment instance created', 'DescribedPaymentInstanceCreated'],
  ['Described asset instance created', 'DescribedAssetInstanceCreated'],
  ['Described payment definition created', 'DescribedPaymentDefinitionCreated'],
  ['Member registered', 'MemberRegistered']
];

const subscriptionInfoCorda = [
  ['Asset instance created', 'io.kaleido.kat.states.AssetInstanceCreated'],
  ['Described asset instance created', 'io.kaleido.kat.states.DescribedAssetInstanceCreated'],
  ['Asset instance batch created', 'io.kaleido.kat.states.AssetInstanceBatchCreated'],
  ['Asset instance property set', 'io.kaleido.kat.states.AssetInstancePropertySet']
]

export const ensureEventStreamAndSubscriptions = async () => {
  if(!config.eventStreams.skipSetup) {
    const esMgr = new EventStreamManager(config);
    await esMgr.ensureEventStreamsWithRetry();
  }
};

class EventStreamManager {
  private gatewayPath: string;
  private api: AxiosInstance;
  private streamDetails: any;
  private retryCount: number;
  private retryDelay: number;
  private protocol: string;

  constructor(config: IConfig) {
    const apiURL = new URL(config.apiGateway.apiEndpoint);
    this.gatewayPath = apiURL.pathname.replace(/^\//, '');
    apiURL.pathname = '';
    const creds = `${config.apiGateway.auth?.user??config.appCredentials.user}:${config.apiGateway.auth?.password??config.appCredentials.password}`;
    this.api = axios.create({
      baseURL: apiURL.href,
      headers: {
        Authorization: `Basic ${Buffer.from(creds).toString('base64')}`
      }
    });
    this.api.interceptors.request.use(requestLogger);
    this.api.interceptors.response.use(responseLogger, errorLogger);
    this.retryCount = 20;
    this.retryDelay = 5000;
    this.protocol = config.protocol;
    this.streamDetails = {
      name: config.eventStreams.topic,
      errorHandling: config.eventStreams.config?.errorHandling??'block',
      batchSize: config.eventStreams.config?.batchSize??50,
      batchTimeoutMS: config.eventStreams.config?.batchTimeoutMS??500,
      type: "websocket",
      websocket: {
        topic: config.eventStreams.topic
      }
    }
    if(this.protocol === 'ethereum') {
      // Due to spell error in ethconnect, we do this for now until ethconnect is updated
      this.streamDetails.blockedReryDelaySec = config.eventStreams.config?.blockedRetryDelaySec??30;
    } else {
      this.streamDetails.blockedRetryDelaySec = config.eventStreams.config?.blockedRetryDelaySec??30;
    }
  }

  async ensureEventStreamsWithRetry() {
    for (let i = 1; i <= this.retryCount; i++) {
      try {
        if (i > 1) await sleep(this.retryDelay);
        const stream: IEventStream = await this.ensureEventStream();
        await this.ensureSubscriptions(stream);
        return;
      } catch (err) {
        logger.error(`Attempt ${i} to initialize event streams failed`, err);
      }
    }
    throw new Error("Failed to initialize event streams after retries");
  }

  async ensureEventStream(): Promise<IEventStream> {
    const { data: existingStreams } = await this.api.get('eventstreams');
    let stream = existingStreams.find((s: any) => s.name === this.streamDetails.name);
    if (stream) {
      const { data: patchedStream } = await this.api.patch(`eventstreams/${stream.id}`, this.streamDetails);
      return patchedStream;
    }
    const { data: newStream } = await this.api.post('eventstreams', this.streamDetails);
    return newStream;
  }

  subscriptionInfo() {
    switch(this.protocol) {
      case 'ethereum':
        return subscriptionsInfoEthereum;
      case 'corda':
        return subscriptionInfoCorda;
      default:
        throw new Error("Unsupported protocol.");
    }
  }

  async createSubscription(eventType: string, streamId: string, description: string): Promise<{id: string, name: string}> {
    switch(this.protocol) {
      case 'ethereum':
        return this.api.post(`${this.gatewayPath}/${eventType}/Subscribe`, {
          description,
          name: eventType,
          stream: streamId,
          fromBlock: "0", // Subscribe from the start of the chain
        }).then(r => { logger.info(`Created subscription ${eventType}: ${r.data.id}`); return r.data });
      case 'corda': 
        return this.api.post('subscriptions', {
          name: eventType,
          stream: streamId,
          fromTime: null, // BEGINNING is specified as `null` in Corda event streams
          filter: {
            stateType: eventType,
            stateStatus: "unconsumed",
            relevancyStatus: "all"
          }
        }).then(r => { logger.info(`Created subscription ${eventType}: ${r.data.id}`); return r.data });
      default:
        throw new Error("Unsupported protocol.");
    }
  }



  async ensureSubscriptions(stream: IEventStream) {
    const dbSubscriptions = (await database.retrieveSubscriptions()) || {};
    const { data: existing } = await this.api.get('subscriptions');
    for (const [description, eventName] of this.subscriptionInfo()) {
      const dbEventName = eventName.replace(/\./g, '_');
      let sub = existing.find((s: IEventStreamSubscription) => s.name === eventName && s.stream === stream.id);
      let storedSubId = dbSubscriptions[dbEventName];
      if (!sub || sub.id !== storedSubId) {
        if (sub) {
          logger.info(`Deleting stale subscription that does not match persisted id ${storedSubId}`, sub);
          await this.api.delete(`subscriptions/${sub.id}`);
        }
        const newSub = await this.createSubscription(eventName, stream.id, description);
        dbSubscriptions[dbEventName] = newSub.id;
      } else {
        logger.info(`Subscription ${eventName}: ${sub.id}`);
      }
    }
    await database.upsertSubscriptions(dbSubscriptions);
  }

}

File: internal/blockchain/cardano/eventstream.go
// Copyright © 2025 IOG Singapore and SundaeSwap, Inc.
//
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cardano

import (
	"context"
	"fmt"

	"github.com/go-resty/resty/v2"
	"github.com/hyperledger/firefly-common/pkg/ffresty"
	"github.com/hyperledger/firefly-common/pkg/log"
	"github.com/hyperledger/firefly/internal/coremsgs"
)

type streamManager struct {
	client       *resty.Client
	batchSize    uint
	batchTimeout int64
}

type eventStream struct {
	ID             string `json:"id"`
	Name           string `json:"name"`
	ErrorHandling  string `json:"errorHandling"`
	BatchSize      uint   `json:"batchSize"`
	BatchTimeoutMS int64  `json:"batchTimeoutMS"`
	Type           string `json:"type"`
	Timestamps     bool   `json:"timestamps"`
}

type listener struct {
	ID   string `json:"id"`
	Name string `json:"name,omitempty"`
}

type filter struct {
	Event eventfilter `json:"event"`
}

type eventfilter struct {
	Contract  string `json:"contract"`
	EventPath string `json:"eventPath"`
}

func newStreamManager(client *resty.Client, batchSize uint, batchTimeout int64) *streamManager {
	return &streamManager{
		client:       client,
		batchSize:    batchSize,
		batchTimeout: batchTimeout,
	}
}

func (s *streamManager) getEventStreams(ctx context.Context) (streams []*eventStream, err error) {
	res, err := s.client.R().
		SetContext(ctx).
		SetResult(&streams).
		Get("/api/v1/eventstreams")
	if err != nil || !res.IsSuccess() {
		return nil, ffresty.WrapRestErr(ctx, res, err, coremsgs.MsgCardanoconnectRESTErr)
	}
	return streams, nil
}

func buildEventStream(topic string, batchSize uint, batchTimeout int64) *eventStream {
	return &eventStream{
		Name:           topic,
		ErrorHandling:  "block",
		BatchSize:      batchSize,
		BatchTimeoutMS: batchTimeout,
		Type:           "websocket",
		Timestamps:     true,
	}
}

func (s *streamManager) createEventStream(ctx context.Context, topic string) (*eventStream, error) {
	stream := buildEventStream(topic, s.batchSize, s.batchTimeout)
	res, err := s.client.R().
		SetContext(ctx).
		SetBody(stream).
		SetResult(stream).
		Post("/api/v1/eventstreams")
	if err != nil || !res.IsSuccess() {
		return nil, ffresty.WrapRestErr(ctx, res, err, coremsgs.MsgCardanoconnectRESTErr)
	}
	return stream, nil
}

func (s *streamManager) updateEventStream(ctx context.Context, topic string, batchSize uint, batchTimeout int64, eventStreamID string) (*eventStream, error) {
	stream := buildEventStream(topic, batchSize, batchTimeout)
	res, err := s.client.R().
		SetContext(ctx).
		SetBody(stream).
		SetResult(stream).
		Patch("/api/v1/eventstreams/" + eventStreamID)
	if err != nil || !res.IsSuccess() {
		return nil, ffresty.WrapRestErr(ctx, res, err, coremsgs.MsgCardanoconnectRESTErr)
	}
	return stream, nil
}

func (s *streamManager) ensureEventStream(ctx context.Context, topic string) (*eventStream, error) {
	existingStreams, err := s.getEventStreams(ctx)
	if err != nil {
		return nil, err
	}
	for _, stream := range existingStreams {
		if stream.Name == topic {
			stream, err = s.updateEventStream(ctx, topic, s.batchSize, s.batchTimeout, stream.ID)
			if err != nil {
				return nil, err
			}
			return stream, nil
		}
	}
	return s.createEventStream(ctx, topic)
}

func (s *streamManager) getListener(ctx context.Context, streamID string, listenerID string, okNotFound bool) (listener *listener, err error) {
	res, err := s.client.R().
		SetContext(ctx).
		SetResult(&listener).
		Get(fmt.Sprintf("/api/v1/eventstreams/%s/listeners/%s", streamID, listenerID))
	if err != nil || !res.IsSuccess() {
		if okNotFound && res.StatusCode() == 404 {
			return nil, nil
		}
		return nil, ffresty.WrapRestErr(ctx, res, err, coremsgs.MsgCardanoconnectRESTErr)
	}
	return listener, nil
}

func (s *streamManager) getListeners(ctx context.Context, streamID string) (listeners *[]listener, err error) {
	res, err := s.client.R().
		SetContext(ctx).
		SetResult(&listeners).
		Get(fmt.Sprintf("/api/v1/eventstreams/%s/listeners", streamID))
	if err != nil || !res.IsSuccess() {
		return nil, ffresty.WrapRestErr(ctx, res, err, coremsgs.MsgCardanoconnectRESTErr)
	}
	return listeners, nil
}

func (s *streamManager) createListener(ctx context.Context, streamID, name, lastEvent string, filters []filter) (listener *listener, err error) {
	body := map[string]interface{}{
		"name":      name,
		"type":      "events",
		"fromBlock": lastEvent,
		"filters":   filters,
	}

	res, err := s.client.R().
		SetContext(ctx).
		SetBody(body).
		SetResult(&listener).
		Post(fmt.Sprintf("/api/v1/eventstreams/%s/listeners", streamID))

	if err != nil || !res.IsSuccess() {
		return nil, ffresty.WrapRestErr(ctx, res, err, coremsgs.MsgCardanoconnectRESTErr)
	}

	return listener, nil
}

func (s *streamManager) deleteListener(ctx context.Context, streamID, listenerID string) error {
	res, err := s.client.R().
		SetContext(ctx).
		Delete(fmt.Sprintf("/api/v1/eventstreams/%s/listeners/%s", streamID, listenerID))

	if err != nil || !res.IsSuccess() {
		return ffresty.WrapRestErr(ctx, res, err, coremsgs.MsgCardanoconnectRESTErr)
	}
	return nil
}

func (s *streamManager) ensureFireFlyListener(ctx context.Context, namespace string, version int, address, firstEvent, streamID string) (l *listener, err error) {
	existingListeners, err := s.getListeners(ctx, streamID)
	if err != nil {
		return nil, err
	}

	name := fmt.Sprintf("%s_%d_BatchPin", namespace, version)
	for _, l := range *existingListeners {
		if l.Name == name {
			return &l, nil
		}
	}

	filters := []filter{{
		eventfilter{
			Contract:  address,
			EventPath: "BatchPin",
		},
	}}
	if l, err = s.createListener(ctx, streamID, name, firstEvent, filters); err != nil {
		return nil, err
	}
	log.L(ctx).Infof("BatchPin subscription: %s", l.ID)
	return l, nil
}
